packages/sync/src/conflicts/conflictLink.ts

```diff
@@ -0,0 +1,51 @@
+import { ApolloLink } from "apollo-link";
+import { onError } from "apollo-link-error";
+import { GraphQLError } from "graphql";
+import { DataSyncConfig } from "../config/DataSyncConfig";
+import { ConflictResolutionData, strategies } from "./strategies";
+
+export class ConflictLink {
+    private config: DataSyncConfig;
+
+    constructor(config: DataSyncConfig) {
+        this.config = config;
+    }
+
+    /**
+     * Initialise the conflict link
+     * @returns An apollo link capable of detecting errors
+     */
+    public init(): ApolloLink {
+        const link = onError(({ graphQLErrors, operation, forward }) => {
+            const data = this.getConflictData(graphQLErrors);
+            if (!this.config.conflictStrategy) {
+                this.config.conflictStrategy = strategies.diffMergeClientWins;
+            }
+            const resolvedConflict = this.config.conflictStrategy(data, operation.variables);
```

Comment:
is this function always a synchronised one? Is it possible to support async ones?

Reply:
@wei-lee It looks like onError doesn't support async functions at the moment. https://github.com/apollographql/apollo-link/issues/646#issuecomment-401399248

Reply:
ok, does this mean it will not be possible to allow developers to implement a strategy where it may involve building some UIs to let user choose?

Reply:
@wei-lee it looks like Apollo relies on Observables to do things like this. I am trying to get it working with these at the moment so I will update once I know for sure.

Reply:
onError do not need to be async. We could process everything in observable/promise that will deliver resolved object and pass it forward into chain.  

On the other note we might also need to support different conflictResolver per type or at least enable devs to handle that case